src/platform/plugins/shared/workflows_execution_engine/server/step/foreach_step/enter_foreach_node_impl.ts

```diff
@@ -98,52 +96,33 @@ export class EnterForeachNodeImpl implements NodeImplementation {
     this.wfExecutionRuntimeManager.navigateToNextNode();
   }
 
-  private getForeachExpression(): string {
-    if (!this.node.configuration.foreach) {
+  private getItems(): unknown[] {
+    const expression = this.node.configuration.foreach;
+
+    if (!expression) {
       throw new Error(
-        `Foreach configuration is required. Please specify an array or expression that evaluates to an array.`
+        'Foreach configuration is required. Please specify an array or expression that evaluates to an array.'
       );
     }
 
-    const renderedForeachExpression =
-      this.stepExecutionRuntime.contextManager.renderValueAccordingToContext(
-        this.node.configuration.foreach
-      );
-
-    return renderedForeachExpression;
-  }
-
-  private getItems(renderedForeachExpression: string): unknown[] {
-    let resolvedValue;
-
-    if (isPropertyAccess(renderedForeachExpression)) {
-      const result =
-        this.stepExecutionRuntime.contextManager.readContextPath(renderedForeachExpression);
+    // Try to resolve the expression as JSON first
+    let resolvedValue = this.tryParseJSON(expression);
```

Comment:
It looks like this part is redundant - for a valid array, it can simply be passed to `evaluateExpressionInContext`, which returns the object as-is. It will then be converted to JSON later if needed. It’s not a bug, just unnecessary code duplication.

Reply:
I think we're trying to get too much from the thing that is not designed for it.

What is supported at the moment is this:
```yaml
foreach: {{ steps.foo.output | split: ',' }}
```
Will return the object that is result of "split" filter. Enables the user to have some basic transformation without JSON.parse/JSON.stringify overhead internally.

```yaml
foreach: '["item1", "item2", "item3"]'
```
The hardcoded items that the foreach will go through.